### ISSUE TITLE
'Delete' key does not have repeat function under Wayland

### ISSUE THREAD

**Issue by nicolamori** [eclipse-platform/eclipse.platform.swt#938]
**Describe the bug**
In a Wayland session, holding the delete key does not produce repeated delete action in any editor. Other keys work fine. In other text element, e.g. search box, the delete key repeat functionality works.

**To Reproduce**
- Start Eclipse in a Wayland environment without setting the `GDK_BACKEND` env variable to `x11`
- Open an editor, write some text, put back the cursor to beginning of text then press and hold the delete key
- Verify that only the first character is deleted

**Expected behavior**
Characters are continued to be deleted as long as the key is held.

**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [ ] All OS
- - [ ] Windows
- - [x] Linux
- - [ ] macOS 

2. Additional OS info (e.g. OS version, Linux Desktop, etc)
ArchLinux with Plasma desktop and Wayland session

3. JRE/JDK version
OpenJDK 21.u35 (ArchLinux package jre-openjdk 21.u35-8 from official repositories)
 
**Version since**
Verified with 4.30.0, first reported on 4.10 (see below)

**Workaround (or) Additional context**
Setting `GDK_BACKEND=x11` fixes the issue
Original bug report: https://bugs.eclipse.org/bugs/show_bug.cgi?id=544518

**Comment by NotAlexNoyle** [comment]
I am having this issue on wayland under void linux on eclipse 2023-09

**Comment by phileimer** [comment]
I confirm this issue on Ubuntu 23.10, Eclipse CDT 2023-12, with OpenJDK 21.0.1+12.
Del key repeat works well in a Xorg session.

This is the last problem preventing me from switching to Wayland.

**Comment by boaks** [comment]
Where should that `GDK_BACKEND=x11` be set?
In `eclipse.ini`? Or where?

**Comment by ChthonVII** [comment]
I'm experiencing the same bug **_under X11_** in dbeaver. (See https://github.com/dbeaver/dbeaver/issues/22581.) Obviously, setting `GDK_BACKEND=x11` doesn't help. Assuming it's the same bug, this bug is not specific to Wayland. The other possibility is two different bugs both breaking the delete key in the same way.

**Comment by padamstx** [comment]
I'm seeing this same issue on Fedora 39 and also saw it on F38 as well.

**Comment by culturedcait** [comment]
I'm observing same bug as described above.
Debian 12 bookworm. Wayland.
Eclipse version: 2024-12 (4.34.0), Build 20241128.

**Comment by sarxos** [comment]
Same issue on Ubuntu 24.

Eclipse:

Version: 2025-12 (4.38.0)
Build id: 20251204-0850

But I can confirm that setting `GDK_BACKEND=x11` env variable before starting Eclipse on Ubuntu 24 solves the problem and I do not observe side effects. It also addressed some other inconveniances I observed (like splash screen being off center, detaching panels from Eclipse views by drag'n'drop being somehow broken).

I created the following shell wrapper and put it in `~/bin`:

```bash
#!/bin/bash

export GDK_BACKEND=x11
cd ~/bin/eclipse-4.38
./eclipse
```

Then added an app menu item with `menulibre`.

**Comment by guss77** [comment]
> Same issue on Ubuntu 24.
> But I can confirm that setting `GDK_BACKEND=x11` env variable before starting Eclipse on Ubuntu 24 solves the problem and I do not observe side effects.

I can confirm that using XWayland to run Eclipse works around the broken Wayland implementation, as it affects the `Delete` key repeat.

It also works around another problem I was experiencing, and wasn't sure if its an SWT issue or not, where trying to use the main menu accelerators (<kbd>ALT</kbd> + letter) always opens the "File" menu, regardless of the letter used. I'll open a separate issue of that.

**Comment by Axidi** [comment]
I believe this is an Eclipse-specific issue.

When I run a small SWT application using Text or StyledText, the DEL key repeat works perfectly. However, in Eclipse itself, editing any text file (Java, .txt, .properties, etc.) the DEL key repeat does not work.

**Comment by AlexGdevGT** [comment]
I have the same issue, recently I upgrade my popOS from 22.04 (Gnome, X11) to 24.04 (Cosmic, wayland) and the DEL key  repeat doesn't work.

**Comment by mtilsted** [comment]
Just wanted to say I have almost the same issue on Fedora 43. 

But my case might be a bit difficult, because for me the delete key just stops working at all. Not even a single delete click works.

Right now I use the workaround
export GDK_BACKEND=x11
To run on x11 but that causes other problems (Wrong window positions, different fonts and so on), so I hope this get's fixed so eclipse works 100% on wayland.

And this issue started when I upgraded Fedora. I have been running eclipse on previous versions of Fedora without any issues with the delete button.

**Comment by guss77** [comment]
> And this issue started when I upgraded Fedora. I have been running eclipse on previous versions of Fedora without any issues with the delete button.

@mtilsted - did you run under Wayland on previous Fedora? Because if this worked well for you on Fedora 42 on Wayland, that may be an important clue as to how to fix the problem.

I suspect that you ran GNOME X11 on Fedora 42, and as Fedora 43 removes the GNOME X11 session, you were moved to Wayland and started encountering this problem.
